File: 2022/12.py
from utils.grid import get_neighbors


def get_input(puzzle: str) -> tuple[tuple[int, int], tuple[int, int], list[list[int]]]:
    grid = []
    start = None
    end = None
    for i, line in enumerate(puzzle.splitlines()):
        grid.append([ord(c) - 97 if c not in "SE" else {"S": 0, "E": 25}[c] for c in line])
        if "S" in line:
            start = line.index("S"), i
        if "E" in line:
            end = line.index("E"), i
    assert start and end
    return start, end, grid


def bfs(start, grid, target, step):
    queue: list[tuple[int, int, int]] = [(0, *start)]
    visited = set()
    while queue:
        d, x, y = queue.pop(0)

        if target(x, y):
            return d

        if (x, y) in visited:
            continue
        visited.add((x, y))

        for p, q in get_neighbors(x, y, len(grid[0]), len(grid)):
            if step(x, y, p, q) and (p, q) not in visited:
                queue.append((d + 1, p, q))


def part1(puzzle: str):
    start, end, grid = get_input(puzzle)
    return bfs(start, grid, lambda x, y: (x, y) == end, lambda x, y, p, q: grid[q][p] - grid[y][x] <= 1)


def part2(puzzle: str):
    _, end, grid = get_input(puzzle)
    return bfs(end, grid, lambda x, y: grid[y][x] == 0, lambda x, y, p, q: grid[y][x] - grid[q][p] <= 1)


if __name__ == "__main__":
    from aoc import run

    run(2022, 12, part1, part2)

File: 2022/12.rs
use std::collections::VecDeque;

#[derive(Clone, PartialEq, Eq)]
struct Position {
    x: usize,
    y: usize,
}

impl Position {
    fn neighbors(&self) -> Vec<Position> {
        [(1, 2), (1, 0), (2, 1), (0, 1)]
            .into_iter()
            .filter(|&(dx, dy)| self.x + dx > 0 && self.y + dy > 0)
            .map(|(dx, dy)| Position {
                x: self.x + dx - 1,
                y: self.y + dy - 1,
            })
            .collect()
    }
}

struct Input {
    start: Position,
    end: Position,
    grid: Vec<Vec<u8>>,
}

fn setup(input: &str) -> Input {
    let mut grid = Vec::new();
    let mut start = None;
    let mut end = None;
    for (i, line) in input.trim().lines().enumerate() {
        let mut row = Vec::new();
        for (j, c) in line.bytes().enumerate() {
            let c = match c {
                b'S' => {
                    start = Some(Position { x: j, y: i });
                    0
                }
                b'E' => {
                    end = Some(Position { x: j, y: i });
                    25
                }
                _ => c - b'a',
            };
            row.push(c);
        }
        grid.push(row);
    }
    Input {
        start: start.unwrap(),
        end: end.unwrap(),
        grid,
    }
}

fn bfs(
    start: Position,
    grid: &[&[u8]],
    target: impl Fn(&Position) -> bool,
    step: impl Fn(&Position, &Position) -> bool,
) -> u32 {
    let mut queue = VecDeque::from([(0, start)]);
    let mut visited = vec![vec![false; grid[0].len()]; grid.len()];
    while let Some((d, p)) = queue.pop_front() {
        if target(&p) {
            return d;
        }
        if visited[p.y][p.x] {
            continue;
        }
        visited[p.y][p.x] = true;
        for q in p.neighbors() {
            if grid.get(q.y).and_then(|r| r.get(q.x)).is_some()
                && step(&p, &q)
                && !visited[q.y][q.x]
            {
                queue.push_back((d + 1, q));
            }
        }
    }
    panic!()
}

fn part1(Input { start, end, grid }: &Input) -> u32 {
    bfs(
        start.clone(),
        &grid.iter().map(|x| &x[..]).collect::<Vec<_>>()[..],
        |pos| pos == end,
        |from, to| grid[to.y][to.x] <= grid[from.y][from.x] + 1,
    )
}

fn part2(Input { end, grid, .. }: &Input) -> u32 {
    bfs(
        end.clone(),
        &grid.iter().map(|x| &x[..]).collect::<Vec<_>>()[..],
        |pos| grid[pos.y][pos.x] == 0,
        |from, to| grid[from.y][from.x] <= grid[to.y][to.x] + 1,
    )
}

aoc::main!(2022, 12);
aoc::example!(ex01, "12.1.txt", 31, 29);
aoc::test_input!("12.txt", 380, 375);
